src/board.ts

```diff
@@ -334,6 +335,155 @@ export function getKeyAtDomPos(pos: cg.NumberPair, asWhite: boolean, bounds: Cli
   return (file >= 0 && file < 8 && rank >= 0 && rank < 8) ? pos2key([file, rank]) : undefined;
 }
 
+function isAlreadySnapped(orig: cg.Pos, targetKey: cg.Key | undefined): boolean {
+  // TODO profile performance - precompute this function's values to an object?
+  if (targetKey === undefined) return false;
+  const pos = key2pos(targetKey);
+  // +
+  if (orig[0] === pos[0] || orig[1] === pos[1]) return true;
+  // x
+  if (Math.abs(orig[0] - pos[0]) === Math.abs(orig[1] - pos[1])) return true;
+  // knight
+  if (Math.abs(orig[0] - pos[0]) === 2 && Math.abs(orig[1] - pos[1]) === 1) return true;
+  if (Math.abs(orig[0] - pos[0]) === 1 && Math.abs(orig[1] - pos[1]) === 2) return true;
+  // all other
+  return false;
+}
+
+export function getSnappedKeyAtDomPos(orig: cg.Key, pos: cg.NumberPair, asWhite: boolean, bounds: ClientRect): cg.Key | undefined {
+  // 1. Get key at dom pos
+  // 2. If move is valid, short circuit
+  // 3. If move is not valid, pick nearest angle and snap distance to dragged distance
+  // Only snapping to + and x currently.
+  // TODO: also snap to knight angles, but stop after going past knight tile?
+  // TODO: cleanly roll these if statements into 1 loop?
+
+  const unsnappedKey = getKeyAtDomPos(pos, asWhite, bounds);
+  const origRowCol = key2pos(orig);
+  if (isAlreadySnapped(origRowCol, unsnappedKey)) return unsnappedKey;
+
+  const draggedOutOfBox = unsnappedKey === undefined;
+  const unsnappedRowCol = unsnappedKey === undefined ? origRowCol : key2pos(unsnappedKey);
+  const origCenterCoord = computeSquareCenter(orig, asWhite, bounds);
+
+  // Use degrees for ease of round numbers in code
+  const origCenterToMouseDegrees = Math.atan2(pos[1] - origCenterCoord[1], origCenterCoord[0] - pos[0]) * 180 / Math.PI + 180;
+  const origCenterToMouseCoordX = origCenterCoord[0] - pos[0];
+  const origCenterToMouseCoordY = origCenterCoord[1] - pos[1];
+
+  const origCenterToMouseCoordDistance = Math.sqrt(
+      origCenterToMouseCoordX * origCenterToMouseCoordX +
+      origCenterToMouseCoordY * origCenterToMouseCoordY
+  );
+  const diagonalCoordDistance = Math.sqrt(
+      bounds.width * bounds.width +
+      bounds.height * bounds.height
+  ) / 8;
+  const squareCoordDistance = Math.floor((origCenterToMouseCoordDistance + diagonalCoordDistance / 2) / diagonalCoordDistance);
+
+  if (origCenterToMouseDegrees < (45 * 0 + 45 / 2)) {
+    // console.log('e+');
+    if (draggedOutOfBox) {
```

Comment:
I put this in to allow dragging a bit off the board, but now I notice:
* when an arrow from a corner (like a1) along the edge and continuing to drag off the board (like imaginary e0 or e-1), the arrow snaps to the full length (like h1)
* when dragging the right mouse button from the board and releasing it off the board, the Firefox context menu appears

Do we want to keep the ability to drag off the board? If yes, any desires on what the correct behavior for issue 1 is?

Reply:
My latest push includes a change to disable dragging an arrow off the board, like the current behavior, just to make things easy for now.